File: backend/src/models/DoctorAvailability.js
const { Model, DataTypes } = require('sequelize');

class DoctorAvailability extends Model {
  static initModel(sequelize) {
    DoctorAvailability.init({

  id: {
    type: DataTypes.INTEGER,
    primaryKey: true,
    autoIncrement: true
  },
  doctorId: {
    type: DataTypes.INTEGER,
    allowNull: false,
    references: {
      model: 'doctors',
      key: 'id'
    }
  },
  dayOfWeek: {
    type: DataTypes.INTEGER, // 0 = Sunday, 1 = Monday, etc.
    allowNull: false,
    validate: {
      min: 0,
      max: 6
    }
  },
  startTime: {
    type: DataTypes.STRING, // Format: "09:00"
    allowNull: false
  },
  endTime: {
    type: DataTypes.STRING, // Format: "17:00"
    allowNull: false
  },
  isAvailable: {
    type: DataTypes.BOOLEAN,
    defaultValue: true
  },
  maxAppointments: {
    type: DataTypes.INTEGER,
    defaultValue: 10
  },
  appointmentDuration: {
    type: DataTypes.INTEGER, // in minutes
    defaultValue: 30
  },
  breakStart: {
    type: DataTypes.STRING, // Format: "12:00"
    allowNull: true
  },
  breakEnd: {
    type: DataTypes.STRING, // Format: "13:00"
    allowNull: true
  }

    }, {
      sequelize,
      modelName: 'DoctorAvailability',

  timestamps: true,
  tableName: 'doctor_availabilities',
  indexes: [
    {
      fields: ['doctorId', 'dayOfWeek']
    }
  ]

    });
    return DoctorAvailability;
  }

  static associate(models) {
    // Associations will be defined in models/index.js
  }
}

module.exports = DoctorAvailability;

File: fastapi_backend/models/doctor_availability.py
from beanie import Document
from pydantic import BaseModel, Field
from typing import Optional
from datetime import datetime

# Pydantic schemas
class DoctorAvailabilityBase(BaseModel):
    doctor_id: str
    day_of_week: int = Field(..., ge=0, le=6)  # 0 = Sunday, 1 = Monday, etc.
    start_time: str  # Format: "09:00"
    end_time: str  # Format: "17:00"
    is_available: bool = True
    max_appointments: int = Field(default=10, ge=1)
    appointment_duration: int = Field(default=30, ge=1)  # in minutes
    break_start: Optional[str] = None  # Format: "12:00"
    break_end: Optional[str] = None  # Format: "13:00"

class DoctorAvailabilityCreate(DoctorAvailabilityBase):
    pass

class DoctorAvailabilityUpdate(BaseModel):
    start_time: Optional[str] = None
    end_time: Optional[str] = None
    is_available: Optional[bool] = None
    max_appointments: Optional[int] = Field(None, ge=1)
    appointment_duration: Optional[int] = Field(None, ge=1)
    break_start: Optional[str] = None
    break_end: Optional[str] = None

class DoctorAvailabilityResponse(DoctorAvailabilityBase):
    id: str
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True

# Beanie Document
class DoctorAvailability(Document, DoctorAvailabilityBase):
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)

    class Settings:
        name = "doctor_availabilities"
        indexes = [
            "doctor_id",
            "day_of_week",
            "is_available",
            "created_at"
        ]

    def __str__(self):
        return f"DoctorAvailability(id={self.id}, doctor_id={self.doctor_id}, day={self.day_of_week})"
